src/lib/tests/visitors/decorators/testsVisitorDecorator.ts

```diff
@@ -0,0 +1,19 @@
+import { TestsVisitor } from '../testVisitor';
+import { Test } from '../../test';
+import { TestSuite } from '../../testSuite';
+
+export abstract class TestsVisitorDecorator<T> implements TestsVisitor<T> {
+    protected baseVisitTest: (testSuite: Test) => Promise<T>;
+    protected baseVisitTestSuite: (testSuite: TestSuite) => Promise<T>;
+
+    constructor(private baseVisitor: TestsVisitor<T>) {
+        this.baseVisitTest = baseVisitor.visitTest.bind(baseVisitor);
+        this.baseVisitTestSuite = baseVisitor.visitTestSuite.bind(baseVisitor);
+
+        this.baseVisitor.visitTest = this.visitTest.bind(this);
+        this.baseVisitor.visitTestSuite = this.visitTestSuite.bind(this);
```

Comment:
This hack is very ugly. This is currently necessary because of the recursive nature of the visitor. When the bottommost visitor in the decorator hierarchy does its recursive call and passes `this` to the `accept` method, we lose the decorators. 

I believe to move the iterating-over-children logic from the visitor to the composite might - perhaps - fix the problem.

Reply:
Okay so moving back the iteration inside the tree creates more problems (aka handling before/after methods execution, filtering the nodes, etc.). The hack doesn't seem that ugly after all. Will continue thinking about this.